### PR TITLE
Ramiro's Debug and profiling notes

### PR DESCRIPTION
**Pull request by serraramiro1**

### CODE REVIEW

File: ramiro_serra/Debug and profiling notes.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+This is how my `send_goal.launch` looked like after the changes I made.
+    
+    <launch>
+        <arg name="x" default="2.0"/>
+        <arg name="y" default="2.0"/>
+        <arg name="Y" default="1.0"/>
+        <arg name="ns" default="create1"/>
+        <node pkg="ca_move_base" type="send_goal" name="navigation_goals"
+                args="$(arg x) $(arg y) $(arg Y)" ns="$(arg ns)" launch-prefix="valgrind --tool=callgrind --callgrind-out-file='callgrind.sendgoal.%p'"/>
```
Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md]: What about this?
```xml
<arg name="debug" default="false"/>
<arg unless="$(arg debug)" name="launch_prefix" value=""/>
<arg          if="$(arg debug)" name="launch_prefix" value="valgrind --tool=callgrind --callgrind-out-file='callgrind.sendgoal.%p'"/>

<node ... launch-prefix="$(arg launch_prefix)"/>
```

It seems a much cleaner way of doing it and it's based on `gazebo_ros`' launch file.

File: ramiro_serra/Debug and profiling notes.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    </launch>
+
+Note i added some default values to the required arguments, only to make the usage of the launchfile more straight-forward.
```
Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md]: Use uppercase i to talk about you.

File: ramiro_serra/Debug and profiling notes.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+We can launch now the file with `roslaunch ca_move_base send_goal.launch`, we can let it run a little and then `C-c` to stop it.
+Valgrind will automatically create a log located in `~/.ros/`, let's check it:
+    
+    cd ~/.ros/
+    ls
+
+Then we can open it with `kcachegrind`, that we installed earlier
+    
+    kcachegrind callgrind.sendgoal.10303
+*The appended number may very with the PID associated.*
```
Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md]: Typo: vary?

File: ramiro_serra/Debug and profiling notes.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        <arg name="x" default="2.0"/>
+        <arg name="y" default="2.0"/>
+        <arg name="Y" default="1.0"/>
+        <arg name="ns" default="create1"/>
+        <node pkg="ca_move_base" type="send_goal" name="navigation_goals"
+                args="$(arg x) $(arg y) $(arg Y)" ns="$(arg ns)" launch-prefix="valgrind --tool=callgrind --callgrind-out-file='callgrind.sendgoal.%p'"/>
+    </launch>
+
+Note i added some default values to the required arguments, only to make the usage of the launchfile more straight-forward.
+We can launch now the file with `roslaunch ca_move_base send_goal.launch`, we can let it run a little and then `C-c` to stop it.
```
Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md]: Silly question: `C-c` means `Ctrl + C`, right?

File: ramiro_serra/Debug and profiling notes.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+Valgrind will automatically create a log located in `~/.ros/`, let's check it:
+    
+    cd ~/.ros/
+    ls
+
+Then we can open it with `kcachegrind`, that we installed earlier
```
Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md]: Ok, so `vallgrind` measures and `kcachegrind` analyzes the information.

File: ramiro_serra/Debug and profiling notes.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    
+    kcachegrind callgrind.sendgoal.10303
+*The appended number may very with the PID associated.*
+
+![kcachegrind](media/kcachegrind.png)
+
+Now a window shows up, with a lot of info of the program, if you want to profundize more on the topic, you can see [here](http://valgrind.org/docs/), and learn how to interpret the data.
```
Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md]: I guess that "profundize" doesn't exist.

Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md]: Could you mention at least one feature that matters?

File: ramiro_serra/Debug and profiling notes.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+Then run the program with `run`.
+We can see the program stopped at the checkpoint we set previously.
+
+    Breakpoint 1, __static_initialization_and_destruction_0 (__initialize_p=1, 
+        __priority=65535)
+        at /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp:10
+    10      const std::string NODE_NAME = "navigation_goals";
+
+We could watch a variable with `watch`, and many more features.
```
Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md:115]: what do you see if you run `where full`?

**Review by eborghi10 — CHANGES_REQUESTED** [review]
I left some comments, PTAL. Well done so far!

File: ramiro_serra/Debug and profiling notes.md (hunk 4 above)
Comment (serraramiro1) [reply ramiro_serra/Debug and profiling notes.md]: That's correct! It's a notation I've seen on the `tmux` documentation

File: ramiro_serra/Debug and profiling notes.md (hunk 3 above)
Comment (serraramiro1) [reply ramiro_serra/Debug and profiling notes.md]: Adressed

File: ramiro_serra/Debug and profiling notes.md (hunk 2 above)
Comment (serraramiro1) [reply ramiro_serra/Debug and profiling notes.md]: Adressed

File: ramiro_serra/Debug and profiling notes.md (hunk 5 above)
Comment (serraramiro1) [reply ramiro_serra/Debug and profiling notes.md]: Added that info

File: ramiro_serra/Debug and profiling notes.md (hunk 6 above)
Comment (serraramiro1) [reply ramiro_serra/Debug and profiling notes.md]: Changed profundize to `deepen`

Comment (serraramiro1) [reply ramiro_serra/Debug and profiling notes.md]: Adressed

File: ramiro_serra/Debug and profiling notes.md (hunk 7 above)
Comment (serraramiro1) [reply ramiro_serra/Debug and profiling notes.md:115]: Added

File: ramiro_serra/Debug and profiling notes.md (hunk 1 above)
Comment (serraramiro1) [reply ramiro_serra/Debug and profiling notes.md]: Adressed

**Comment by serraramiro1** [comment]
## All changes adressed, please review @eborghi10

File: ramiro_serra/Debug and profiling notes.md [hunk 8]
```diff
@@ -82,6 +84,30 @@ We can see the program stopped at the checkpoint we set previously.
         at /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp:10
     10      const std::string NODE_NAME = "navigation_goals";
 
+Now if we use `where full`:
```
Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md:91]: Do you notice something useful there?

File: ramiro_serra/Debug and profiling notes.md [hunk 9]
```diff
@@ -27,21 +27,24 @@ This is how my `send_goal.launch` looked like after the changes I made.
...
     
     kcachegrind callgrind.sendgoal.10303
-*The appended number may very with the PID associated.*
+*The appended number may vary with the PID associated.*
 
 ![kcachegrind](media/kcachegrind.png)
 
-Now a window shows up, with a lot of info of the program, if you want to profundize more on the topic, you can see [here](http://valgrind.org/docs/), and learn how to interpret the data.
+Now a window shows up, with a lot of info of the program, if you want to deepen more on the topic, you can see [here](http://valgrind.org/docs/), and learn how to interpret the data.
+For example, the camp `Self` means the time that the program was in the function itself, and the camp `Called` measures the time that the program spent on the functions that were called inside the function.
```
Comment (eborghi10) [inline ramiro_serra/Debug and profiling notes.md]: "camp" or "field"?

**Review by eborghi10 — CHANGES_REQUESTED** [review]
One minor change, PTAL.

File: ramiro_serra/Debug and profiling notes.md (hunk 8 above)
Comment (serraramiro1) [reply ramiro_serra/Debug and profiling notes.md:91]: Maybe that when the variables are optimized out, you can't see their value?

**Review by eborghi10 — APPROVED** [review]
LGTM

**Comment by eborghi10** [comment]
Could you put a picture with the launch structure?

**Comment by serraramiro1** [comment]
> Could you put a picture with the launch structure?

Added.
## Ready to merge

**Comment by serraramiro1** [comment]
@eborghi10 
## Ready to merge

**Pull request merged**